### ISSUE TITLE
Contact search menu doesnt update when displayname in LDAP settings is changed

### ISSUE THREAD

**Issue by johkoenig** [nextcloud/server#21517]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Steps to reproduce
1. Set up a NC 19.0.0
2. Set up LDAP. On the tab "Advanced" -> "Directory Settings" write down the following
"User Display Name Field" -> "displayname"
"2nd User Display Name Field" -> "uid"
3. Let everything run for a couple of days. The contact search menu will show up correctly with every user in the style "John Doe (jdoe)". 
4. Change the LDAP settings. From now on, the "2nd Display Name Field" is empty.

### Expected behaviour
The new display names will show up everywhere (userlist, talk, file sharing, ...)

### Actual behaviour
The new user names show up almost everywhere, but not on the contact search menu:
**LDAP Settings:**
![grafik](https://user-images.githubusercontent.com/42836313/85232153-4db49700-b3fd-11ea-942f-da104c9ecf03.png)
**Example for correct behaviour (nextcloud talk):**
![grafik](https://user-images.githubusercontent.com/42836313/85232208-80f72600-b3fd-11ea-8f2f-e5df5957d392.png)
**Incorrect behaviour on the contact search menu:**
![grafik](https://user-images.githubusercontent.com/42836313/85232145-32e22280-b3fd-11ea-9e2f-8a394221fd14.png)

### Server configuration
**Nextcloud version:** 19.0.0.2
**Updated from an older Nextcloud/ownCloud or fresh install:** fresh
**Where did you install Nextcloud from:** nextcloud.com

Note: Contacts app is deactivated

@nextcloud/ldap 
@nextcloud/contacts

**Comment by skjnldsv** [comment]
@blizzz

**Comment by blizzz** [comment]
I cannot reproduce this behaviour with current stable19 branch. There is also no changed code in this regard. The contacts menu is fed from an internal address book, which again is updated when relevant user data was changed, like the display name. What might be is that the entry in the addressbook is dangling, i.e. was not recognized with  the user given on the change. This should not be the case either provided the reproduction steps are complete.

You can run `php occ dav:sync-system-addressbook` to update the addressbook.

**Comment by johkoenig** [comment]
@blizzz 
the occ command you've provided didn't help. The menu still looks the same. However, I think we're nailing down the problem.

What I did, was: 
I changed the LDAP settings from 
```
| ldapUserDisplayName           | displayname                                                                                                                         |
| ldapUserDisplayName2          | uid                                                                                                                                 |         
```
to
```
| ldapUserDisplayName           | displayname                                                                                                                         |
| ldapUserDisplayName2          |                                                                                                                                     |         
```
The LDAP data, such as displayName and uid, remained unchanged. What I expected was that the contact menu changes from ```%ldapUserDisplayName (%ldapUserDisplayName2)``` to ```%ldapUserDisplayName```. However, this didn't happen. Probably due to what @blizzz said that "The contacts menu is fed from an internal address book, which again is updated _when relevant user data was changed_, like the display name." As long as the user data remains unchanged, the displayname for the contacts menu will not be updated. 
You can see this very clearly since a couple of days ago I added another user:
![grafik](https://user-images.githubusercontent.com/42836313/88209407-f64f5400-cc52-11ea-93be-27b269ee6d65.png)
User 1 and 3 (SK and MK) were present before the LDAP settings were changed. User 2 (IK) afterwards.

So, what I propose is to trigger a rebuild of the contacts menu display names as soon as the attributes for the ldapUserDisplayName or ldapUserDisplayName2 is changed/removed.

**Comment by blizzz** [comment]
Changing the settings clears the LDAP cache. As soon as an operation fetches the user from the LDAP server,  the display name will be updated and synced to the address book. Also there is a background job that attempts  to update each user twice a day, this would also update the entry. It is not real time, sure. You could argue whether it is reasonable that the display name change forces a resync of all LDAP users in the system address book. Especially with large instances, i do not think it is useful and would overlap with standard update procedures anyways.

**Comment by johkoenig** [comment]
Based on what you said, I changed the ldap settings again. I set "ldapUserDisplayName2" to "uid" and a minute later I changed it back to empty. On both changes, the names in "talk" and on the contact menu were updated correctly. I don't know whether this has something to do with the update to 19.0.1 which I did a couple of days ago, however, it seems like the bug is not present anymore.

Since this behaviour is not reproducable on my current NC 19.0.1 and everything works as expected, I'm closing this ticket.
